deployment/migrations/config_updater.py

```diff
@@ -0,0 +1,103 @@
+"""
+Checks and updates the configuration and key files of an Aleph node.
+
+This script imports the migration scripts in the scripts/ directory one by one
+and runs the `upgrade` function they must all implement. Alternatively,
+it can revert operations using the `downgrade` function, if implemented.
+
+Migration scripts must implement a check to verify if they need to run
+as the tool itself has no way to determine which scripts have already
+been executed in the past.
+"""
+
+import argparse
+import importlib.util
+import logging
+import os
+import sys
+
+LOGGER = logging.getLogger()
+
+
+SERIALIZED_KEY_FILE = "serialized-node-secret.key"
+
+
+def cli_parse() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(
+        description="Checks and updates the configuration and key files of an Aleph node."
+    )
+    parser.add_argument(
+        "command",
+        action="store",
+        choices=("upgrade", "downgrade"),
+        type=str,
+        help="Path to setup.cfg.",
+    )
+    parser.add_argument(
+        "--key-dir",
+        action="store",
+        required=True,
+        type=str,
+        help="Path to the key directory.",
+    )
+    parser.add_argument(
+        "--key-file",
+        action="store",
+        required=False,
+        type=str,
+        help="Path to the private key file, if any. Only used to upgrade the key to the latest format.",
+    )
+    parser.add_argument(
+        "--verbose",
+        "-v",
+        help="Show more information.",
+        action="store_true",
+        default=False,
+    )
+    return parser.parse_args()
+
+
+def setup_logging(log_level: int) -> None:
+    logformat = "%(asctime)s [%(levelname)s] %(name)s: %(message)s"
+    logging.basicConfig(
+        level=log_level,
+        stream=sys.stdout,
+        format=logformat,
+        datefmt="%Y-%m-%d %H:%M:%S",
+    )
+
+
+def main(args: argparse.Namespace):
+    log_level = logging.DEBUG if args.verbose else logging.INFO
+    setup_logging(log_level)
+
+    migration_scripts_dir = os.path.join(os.path.dirname(__file__), "scripts")
+    migration_scripts = sorted(
+        f for f in os.listdir(migration_scripts_dir) if f.endswith(".py")
+    )
+
+    command = args.command
+
+    for migration_script in migration_scripts:
+        migration_script_path = os.path.join(migration_scripts_dir, migration_script)
+        spec = importlib.util.spec_from_file_location(
```

Comment:
This approach seems pretty complex. Can you explain why such a complex framework instead of either importing and running every migration (should be able to self-check and run more than once) or running executable scripts independently of the language ?

Reply:
I didn't invent anything, this is really similar to the way alembic (the SQLAlchemy migration tool) works. I like it so I figured I'd do something similar (upgrade/downgrade functions in Python scripts, automatic import from a scripts directory).

To answer your questions directly:
* I don't see your point regarding self-check and running more than once? This is up to the migration script. For example, the current script checks that the file to create already exists.
* Language independence: that's a fair point, but worst case just spawn a subprocess in the migration script and that's all.

As for the benefits:
* Automatic discovery of the scripts, impossible to forget one somewhere
* Pretty natural to write, just need to implement upgrade and downgrade and you can forget about all the boilerplate like command-line parsing etc.